chain/state/statetree.go

```diff
@@ -26,16 +26,71 @@ type StateTree struct {
 	root  *hamt.Node
 	Store cbor.IpldStore
 
-	// Maps ID addresses to actors.
-	actorcache map[address.Address]*types.Actor
-	snapshots  []cid.Cid
+	// Maps ID addresses to actors, layered by snapshots
+	snaps *stateSnaps
+}
+
+type stateSnaps struct {
+	layers []map[address.Address]streeOp
+}
+
+type streeOp struct {
+	Act    types.Actor
+	Delete bool
+}
+
+func newStateSnaps() *stateSnaps {
+	return &stateSnaps{
+		layers: []map[address.Address]streeOp{make(map[address.Address]streeOp)},
+	}
+}
+
+func (ss *stateSnaps) addLayer() {
+	ss.layers = append(ss.layers, make(map[address.Address]streeOp))
+}
+
+func (ss *stateSnaps) dropLayer() {
+	ss.layers = ss.layers[:len(ss.layers)-1]
+}
+
+func (ss *stateSnaps) mergeLastLayer() {
+	last := ss.layers[len(ss.layers)-1]
+	nextLast := ss.layers[len(ss.layers)-2]
+
+	for k, v := range last {
+		nextLast[k] = v
+	}
+
+	ss.dropLayer()
```

Comment:
Probably warrants benchmarking - Deleting keys as we merge, then keeping the map instance could save us some allocations

Reply:
Yeah, that seems like an optimization we can do later when it comes down to it.